File: jdroid-android-firebase-admob/src/main/java/com/jdroid/android/firebase/admob/helpers/InterstitialAdHelper.java
package com.jdroid.android.firebase.admob.helpers;

import android.app.Activity;
import android.view.ViewGroup;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.jdroid.android.firebase.admob.AdMobAppModule;
import com.jdroid.android.firebase.admob.AdsStats;
import com.jdroid.android.location.LocationHelper;
import com.jdroid.android.utils.AppUtils;
import com.jdroid.java.exception.UnexpectedException;

import java.util.List;

public class InterstitialAdHelper implements AdHelper {

	private static final String TEST_AD_UNIT_ID = "ca-app-pub-3940256099942544/1033173712";

	private InterstitialAd interstitial;
	private Boolean displayInterstitial = false;
	private String interstitialAdUnitId;
	private List<AdListener> adListeners;

	public InterstitialAdHelper() {
		interstitialAdUnitId = AdMobAppModule.getAdMobAppContext().getDefaultAdUnitId();
	}

	private AdRequest.Builder createBuilder() {
		final AdRequest.Builder builder = new AdRequest.Builder();
		if (!AppUtils.INSTANCE.isReleaseBuildType()) {
			builder.addTestDevice(AdRequest.DEVICE_ID_EMULATOR);
			for (String deviceId : AdMobAppModule.getAdMobAppContext().getTestDevicesIds()) {
				builder.addTestDevice(deviceId);
			}
		}
		builder.setLocation(LocationHelper.get().getLocation());
		return builder;
	}

	@Override
	public void loadAd(Activity activity, ViewGroup adViewContainer) {
		if (AdMobAppModule.getAdMobAppContext().isInterstitialEnabled()) {
			interstitial = new InterstitialAd(activity);

			if (interstitialAdUnitId == null) {
				throw new UnexpectedException("Missing interstitial ad unit ID");
			}

			if (!AppUtils.INSTANCE.isReleaseBuildType() && AdMobAppModule.getAdMobAppContext().isTestAdUnitIdEnabled()) {
				interstitial.setAdUnitId(TEST_AD_UNIT_ID);
			} else {
				interstitial.setAdUnitId(interstitialAdUnitId);
			}

			AdRequest.Builder builder = createBuilder();
			interstitial.loadAd(builder.build());

			AdListenerWrapper adListenerWrapper = new AdListenerWrapper();
			adListenerWrapper.addAdListener(new AdListener() {

				@Override
				public void onAdLoaded() {
					if (displayInterstitial) {
						displayInterstitial(false);
					}
				}

				@Override
				public void onAdOpened() {
					AdsStats.INSTANCE.onInterstitialOpened();
				}
			});
			if (adListeners != null) {
				for (AdListener adListener : adListeners) {
					adListenerWrapper.addAdListener(adListener);
				}
			}
			interstitial.setAdListener(adListenerWrapper);
		}
	}

	public Boolean displayInterstitial() {
		return displayInterstitial(false);
	}

	public Boolean displayInterstitial(Boolean retryIfNotLoaded) {
		return displayInterstitial(retryIfNotLoaded, null);
	}

	public Boolean displayInterstitial(Boolean retryIfNotLoaded, AdListener adListener) {
		displayInterstitial = retryIfNotLoaded;
		if ((interstitial != null) && interstitial.isLoaded()) {
			if (adListener != null) {
				((AdListenerWrapper)interstitial.getAdListener()).addAdListener(adListener);
			}
			interstitial.show();
			displayInterstitial = false;
			return true;
		}
		return false;
	}

	@Override
	public AdHelper setAdUnitId(String adUnitId) {
		this.interstitialAdUnitId = adUnitId;
		return this;
	}

	@Override
	public AdHelper setAdListeners(List<AdListener> adListeners) {
		this.adListeners = adListeners;
		return this;
	}
}

File: jdroid-android-firebase-admob/src/main/java/com/jdroid/android/firebase/admob/helpers/InterstitialAdHelper.kt
package com.jdroid.android.firebase.admob.helpers

import android.app.Activity
import android.view.ViewGroup

import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.InterstitialAd
import com.jdroid.android.firebase.admob.AdMobAppModule
import com.jdroid.android.firebase.admob.AdsStats
import com.jdroid.android.location.LocationHelper
import com.jdroid.android.utils.AppUtils
import com.jdroid.java.exception.UnexpectedException

class InterstitialAdHelper : AdHelper {

    private var interstitial: InterstitialAd? = null
    private var displayInterstitial: Boolean = false
    private var interstitialAdUnitId: String? = null
    private var adListeners: List<AdListener>? = null

    init {
        interstitialAdUnitId = AdMobAppModule.adMobAppContext.getDefaultAdUnitId()
    }

    private fun createBuilder(): AdRequest.Builder {
        val builder = AdRequest.Builder()
        if (!AppUtils.isReleaseBuildType()) {
            builder.addTestDevice(AdRequest.DEVICE_ID_EMULATOR)
            for (deviceId in AdMobAppModule.adMobAppContext.getTestDevicesIds()) {
                builder.addTestDevice(deviceId)
            }
        }
        builder.setLocation(LocationHelper.get().location)
        return builder
    }

    override fun loadAd(activity: Activity, adViewContainer: ViewGroup?) {
        if (AdMobAppModule.adMobAppContext.isInterstitialEnabled()) {
            interstitial = InterstitialAd(activity)

            if (interstitialAdUnitId == null) {
                throw UnexpectedException("Missing interstitial ad unit ID")
            }

            if (!AppUtils.isReleaseBuildType() && AdMobAppModule.adMobAppContext.isTestAdUnitIdEnabled()) {
                interstitial!!.adUnitId = TEST_AD_UNIT_ID
            } else {
                interstitial!!.adUnitId = interstitialAdUnitId
            }

            val builder = createBuilder()
            interstitial!!.loadAd(builder.build())

            val adListenerWrapper = AdListenerWrapper()
            adListenerWrapper.addAdListener(object : AdListener() {

                override fun onAdLoaded() {
                    if (displayInterstitial) {
                        displayInterstitial(false)
                    }
                }

                override fun onAdOpened() {
                    AdsStats.onInterstitialOpened()
                }
            })
            for (adListener in adListeners.orEmpty()) {
                adListenerWrapper.addAdListener(adListener)
            }
            interstitial!!.adListener = adListenerWrapper
        }
    }

    @JvmOverloads
    fun displayInterstitial(retryIfNotLoaded: Boolean = false, adListener: AdListener? = null): Boolean {
        displayInterstitial = retryIfNotLoaded
        if (interstitial != null && interstitial!!.isLoaded) {
            if (adListener != null) {
                (interstitial!!.adListener as AdListenerWrapper).addAdListener(adListener)
            }
            interstitial!!.show()
            displayInterstitial = false
            return true
        }
        return false
    }

    override fun setAdUnitId(adUnitId: String): AdHelper {
        this.interstitialAdUnitId = adUnitId
        return this
    }

    override fun setAdListeners(adListeners: List<AdListener>): AdHelper {
        this.adListeners = adListeners
        return this
    }

    companion object {
        private const val TEST_AD_UNIT_ID = "ca-app-pub-3940256099942544/1033173712"
    }
}
